x/evidence/keeper/abci.go

```diff
@@ -23,7 +22,7 @@ func BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock, k keeper.Keeper)
 		// premeditation. So for now we agree to treat them in the same way.
 		case abci.MisbehaviorType_DUPLICATE_VOTE, abci.MisbehaviorType_LIGHT_CLIENT_ATTACK:
 			evidence := types.FromABCIEvidence(tmEvidence)
-			k.HandleEquivocationEvidence(ctx, evidence.(*types.Equivocation))
```

Comment:
I wouldn't use same evidence interface for `req.ByzantineValidators` as this module requires from external messages. This was the original confusion that left the interfaces such that one could register an evidence handler that blindly trusted `types.Equivocation`.

I would go a bit further.
1. `types.Equivocation` should not implement `Evidence` interface (those must be defined with full proofs and registered)
2. `types.FromABCIEvidence()` should return a concrete type `types.Equivocation` as long as that is the only type we get from Tendermint. If it adds more types later, the module can generalize this.

My key point is [`Evidence`](https://github.com/cosmos/cosmos-sdk/blob/main/x/evidence/exported/evidence.go#L9-L36) should only be things that include their own proof, that can be [validated by a registered handler](https://github.com/cosmos/cosmos-sdk/blob/main/x/evidence/keeper/keeper.go#L82-L92). Allowing an unprovable type as Evidence is a misuse of the type system, declaring something as fitting a hole it must never enter.

Reply:
i went with number two otherwise i would have to introduce more boilerplate for sets since we set Evidence in the store. I think this is safe enough for now. I would prefer we remove this module entirely as i dont see other modules using evidence and instead handling it themself